### PR TITLE
RHOAIENG-1620: [Bug] Asterisk in the pipeline node Output files gets encoded to %2A

### PR DESCRIPTION
**Pull request by hmtosi**
See [RHOAIENG-1620](https://redhat.atlassian.net/browse/RHOAIENG-1620)

## Summary by CodeRabbit

* **New Features**
  * URL decoding for dependency and output paths so encoded wildcards and separators are handled correctly.
  * Path validation to prevent directory traversal while preserving wildcard pattern support.

* **Tests**
  * Added integration and unit tests for URL-decoded wildcard outputs.
  * Expanded tests covering path validation and decoding behavior for dependencies and outputs.

### CODE REVIEW

File: elyra/airflow/bootstrapper.py [hunk 1]
```diff
@@ -150,7 +151,10 @@ def process_outputs(self) -> None:
         if outputs:
             output_list = outputs.split(INOUT_SEPARATOR)
             for file in output_list:
-                self.process_output_file(file.strip())
+                # URL-decode the output file path to handle wildcard characters
+                # that may have been encoded (e.g., * -> %2A)
+                decoded_file = unquote(file.strip())
+                self.process_output_file(decoded_file)
```
[the remark this answers is not in the document]
Comment (hmtosi) [reply elyra/airflow/bootstrapper.py]: This has been addressed, please review again

File: elyra/kfp/bootstrapper.py [hunk 2]
```diff
@@ -162,7 +163,10 @@ def process_outputs(self) -> None:
         if outputs:
             output_list = outputs.split(INOUT_SEPARATOR)
             for file in output_list:
-                self.process_output_file(file.strip())
+                # URL-decode the output file path to handle wildcard characters
+                # that may have been encoded (e.g., * -> %2A)
+                decoded_file = unquote(file.strip())
+                self.process_output_file(decoded_file)
```
[the remark this answers is not in the document]
Comment (hmtosi) [reply elyra/kfp/bootstrapper.py]: This has been addressed, please review again

File: elyra/airflow/bootstrapper.py [hunk 3]
```diff
@@ -130,7 +131,9 @@ def process_dependencies(self) -> None:
         if inputs:
             input_list = inputs.split(INOUT_SEPARATOR)
             for file in input_list:
-                self.get_file_from_object_storage(file.strip())
+                # URL-decode the input file path to handle encoded characters
+                decoded_file = unquote(file.strip())
+                self.get_file_from_object_storage(decoded_file)
```
[the remark this answers is not in the document]
Comment (hmtosi) [reply elyra/airflow/bootstrapper.py:136]: Claude suggests we should stop here (and not implement more edge-case security fixes) for these reasons:

  1. ✅ Original bug is fixed: URL-encoded wildcards now work
  2. ✅ Basic security added: Simple path traversal patterns are blocked
  3. ⚠️  Diminishing returns: More complex attacks (symlinks, etc.) are edge cases in this trusted execution environment
  4. 🔄 Scope creep: We're now doing general security hardening, not bug fixing

File: elyra/tests/kfp/test_url_decode_and_validation.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+        with tempfile.TemporaryDirectory() as tmpdir:
+            original_cwd = os.getcwd()
+            os.chdir(tmpdir)
+            try:
+                safe_paths = ["model/output.txt", "data/results.csv", "subdir/file.bin"]
+
+                for path in safe_paths:
+                    result = kfp_bootstrapper.FileOpBase.validate_path(mock_instance, path)
+                    assert result == path
```
[the remark this answers is not in the document]
Comment (hmtosi) [reply elyra/tests/kfp/test_url_decode_and_validation.py:47]: Claude responds that...

Mock Issue with has_wildcard
✅ Valid concern - The mock prevents testing the non-wildcard branch
✅ Should fix - Use real implementation
Test URL-encoded traversal in validate_path
❌ Going in circles - This tests a scenario that NEVER happens in our code
❌ Misunderstands architecture - Decoding happens BEFORE validation
❌ Not a real vulnerability - You can't pass encoded strings to validate_path in actual usage

[the remark this answers is not in the document]
Comment (hmtosi) [reply elyra/tests/kfp/test_url_decode_and_validation.py:47]: I addressed the first issue with Claude, but not the second. see reasoning below:
  1. Mock Issue with has_wildcard

  - ✅ Valid concern - The mock prevents testing the non-wildcard branch
  - ✅ Should fix - Use real implementation

  2. Test URL-encoded traversal in validate_path

  - ❌ Going in circles - This tests a scenario that NEVER happens in our code
  - ❌ Misunderstands architecture - Decoding happens BEFORE validation
  - ❌ Not a real vulnerability - You can't pass encoded strings to validate_path in actual usage

**Review by jesuino — APPROVED** [review]
I did check the functionality and it is working well. I am running Elyra agains a local KFP installation. Here are my steps:

* Created a script that writes a file to the directory "model". Code:

```
import os
dir = 'model'
if not os.path.exists(dir):
    os.makedirs(dir)
with open(f'./{dir}/test.txt', "w") as f:
    f.write("Hello, world! This is a test file.")
```
* Dragged my notebook to an Elyra pipeline and set the outputs to "model/*". Screenshot:

Image

* When I run the pipeline I can see the output file on my S3 (seaweedFS). 

Image

Repeating these same steps with an elyra built from main I can see that it also works, so after all could this be actually an issue on Data Science Pipelines? If this is the case we can grab the file generated by Elyra and try to upload on KFP/Data Science Pipelines

**Comment by jesuino** [comment]
I did check the issue again on main and could not reproduce it.  I vote to close this PR without merging and later we can revisit it if needed.

**Comment by hmtosi** [comment]
We believe this issue was fixed by the upgrade to `kfp v2.0`. The initial issue was reported on `kfp==1.12.1`. As we are no longer able to reproduce it, we will close the jira as well.
